src/Http/Controllers/WebhookController.php

```diff
@@ -55,6 +56,149 @@ protected function handleCustomerSubscriptionDeleted(array $payload)
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    protected function handleCustomerSubscriptionUpdated(array $payload)
+    {
+        /* @var $user \App\User */
+        $user = $this->getUserByStripeId($payload['data']['object']['customer']);
+
+        if ($user) {
+            $data = $payload['data']['object'];
+
+            $user->subscriptions->filter(function ($subscription) use ($data) {
+                return $subscription->stripe_id === $data['id'];
+            })->each(function ($subscription) use ($data) {
+                /* @var $subscription \Laravel\Cashier\Subscription */
+
+                // Quantity
+                if (isset($data['quantity'])) {
+                    $subscription->quantity = $data['quantity'];
+                }
+
+                // Plan
+                if (isset($data['plan']['id'])) {
+                    $subscription->stripe_plan = $data['plan']['id'];
+                }
+
+                // Trial ends
+                if (isset($data['trial_end'])) {
+                    $trial_ends = Carbon::createFromTimestamp($data['trial_end']);
+                    if ($subscription->trial_ends_at->ne($trial_ends)) {
+                        $subscription->trial_ends_at = $trial_ends;
+                    }
+                }
+
+                // Cancellation
+                if (isset($data['cancel_at_period_end']) && $data['cancel_at_period_end']) {
+                    if ($subscription->onTrial()) {
+                        $subscription->ends_at = $subscription->trial_ends_at;
+                    } else {
+                        $subscription->ends_at = Carbon::createFromTimestamp($data['current_period_end']);
+                    }
+                }
+
+                $subscription->save();
+            });
+        }
+
+        return new Response('Webhook Handled', 200);
+    }
+
+    /**
+     * Handle customer updated.
+     *
+     * @param  array $payload
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    protected function handleCustomerUpdated(array $payload)
+    {
+        /* @var $user \App\User */
+        $user = $this->getUserByStripeId($payload['data']['object']['id']);
+
+        if ($user) {
+            $data = $payload['data']['object'];
+
+            // Change card details
+            if (isset(
+                $data['default_source'],
+                $payload['data']['previous_attributes']['default_source'],
+                $data['sources']['data']
+            )) {
+                $default_card = $data['default_source'];
+                foreach ($data['sources']['data'] as $card) {
+                    if (! isset($card['id']) || $card['id'] != $default_card) {
+                        continue;
+                    }
+                    $user->forceFill([
+                        'card_brand'     => $card['brand'] ?? null,
+                        'card_last_four' => $card['last4'] ?? null,
+                    ])->save();
```

Comment:
Isn't it best to use the `updateCardFromStripe` method here? I'm actually not sure myself if it's best to use that method or to rely on the data of the webhook.

Reply:
I'm not sure too. I don't think it's very useful to make an extra call to Stripe here but maybe I'm wrong.